mlir/lib/Dialect/Rock/Transforms/Regularize.cpp

```diff
@@ -147,187 +146,452 @@ struct RegularizeGenericRewritePattern
-        }
+static void annotateGenericOp(linalg::GenericOp lgop) {
+  MLIRContext *ctx = lgop.getContext();
+  int64_t majorTensorSize = 0;
+  size_t majorTensorIdx;
+  size_t argIdx = -1;
+  if (lgop.getInputs().size() == 1) {
+    lgop->setAttr("rock.majorTensorNumber",
+                  IntegerAttr::get(IndexType::get(ctx), 1));
+    return;
+  }
+  for (auto [inputIdx, inp] : llvm::enumerate(lgop.getInputs())) {
+    while (auto viewOp =
+               dyn_cast_or_null<ViewLikeOpInterface>(inp.getDefiningOp()))
+      inp = viewOp.getViewSource();
+
+    if (isa<BlockArgument>(inp)) {
+      auto arg = dyn_cast<BlockArgument>(inp);
+      auto shape = cast<ShapedType>(inp.getType());
+      int64_t argSize = shape.getNumElements();
+      if (inputIdx == 0 || argSize > majorTensorSize ||
+          (argSize == majorTensorSize && argIdx > arg.getArgNumber())) {
+        majorTensorIdx = inputIdx;
+        majorTensorSize = argSize;
+        argIdx = arg.getArgNumber();
       }
-      inputIdx++;
     }
-    if (majorTensorIdx >= 0)
-      lgop->setAttr("rock.majorTensorNumber",
-                    IntegerAttr::get(IndexType::get(ctx), majorTensorIdx));
   }
-  return;
+  if (majorTensorIdx >= 0)
+    lgop->setAttr("rock.majorTensorNumber",
+                  IntegerAttr::get(IndexType::get(ctx), majorTensorIdx));
 }
 
 ////////////////////////////////////////////////////////////////////////
-////  Push Transforms Over alloc to writer
+////  Classify buffer writes into input and output fusion chains
 ////////////////////////////////////////////////////////////////////////
-struct PushTransformsUpRewritePattern
-    : public OpRewritePattern<memref::AllocOp> {
-  using OpRewritePattern<memref::AllocOp>::OpRewritePattern;
-
-  /////////////////////////////////////////////////////////////////////
-  static bool isFusorOp(Operation *useOp, Value viewedBuffer) {
-    if (auto reduceOp = dyn_cast<rock::ReduceOp>(useOp))
-      return reduceOp.getOut() == viewedBuffer;
-    return isa<rock::GridwiseGemmOp, rock::GridwiseGemmAccelOp,
-               rock::GridwiseAttentionAccelOp, rock::ThreadwiseWriteAllOp>(
-        useOp);
-  }
 
-  static bool collectChain(Value result, Operation *forwOp,
-                           SmallVector<Operation *> &chain) {
-    while (auto top = dyn_cast<rock::TransformOp>(forwOp)) {
-      result = top.getResult();
-      if (!result.hasOneUse()) {
-        // currently restricted to 1 reader
-        LLVM_DEBUG(llvm::dbgs() << "multiple readers on transform\n");
-        return false; // TODO: fix when encountered
-      }
-      chain.push_back(forwOp);
-      forwOp = (*result.getUses().begin()).getOwner();
+namespace {
+struct TransformPushState {
+  llvm::SmallVector<memref::AllocOp> worklist;
+  llvm::SmallPtrSet<OpOperand *, 8> inputFusionWrites;
+  llvm::SmallPtrSet<OpOperand *, 8> outputFusionWrites;
+};
+} // end namespace
+
+/// Given an operand that reads a buffer, collect the writer of thet buffer and,
```

Comment:
What does an operand that reads a buffer mean here ?

In bufferized IR, we have ops with OpOperands that are usually buffers. Therefore I struggle follow this docs what does a operand that reads a buffer mean ?

Reply:
"Operand that reads a buffer" is the operand (Value + place in arg list + operation that's using it) that refers to a buffer that it'll read through some number of views

Reply:
What Im saying is operand itself is the buffer (or a view of it) and its not a reader or writer to the buffer itself.